### PR TITLE
MV-3

### LINKED ISSUE #656: SpeedForwarder (Read this)
<div align="center">
<h1>  </h1>
<p> <h3> <strong> SpeedForwarder: A new Fast Forward structure </strong> </h3> </p>
<p align="center"><a href="https://nightly.link/FastForwardTeam/FastForward/workflows/main/manifest-v3"> GitHub Workflow Status (with branch)</a> </p>
</div>

****

# A new way to bypass
We, the developers of Fast Forward are working on a new update, and with this update, writing bypasses will be easier than never! And also, the so wanted whitelist feature will be added. Check back this issue to see any updates.

_During that big update, we won't take any new bypass requests, only fix broken bypasses._

## Milestone
Currently, our milestone is:

- [x] Redesign how the import_script is handled
- [x] Implement the new project structure
- [x] Add the whitelist feature
- [x] Test it both on Firefox and Chrome browsers
- [ ] See if theres any broken bypass due of the changes
- [ ] Update it on all 3 stores

### PR DESCRIPTION
**Pull request by NotAProton**
tl;dr:
```diff
- "manifest_version": 2,
+ "manifest_version": 3,
```
---

- Added eslint and prettier
- Implemented whitelist
- Changed how injection script and content script communicate
- Removed preflight bypasses
- Removed referrer cache and bypass_clipboard
- Removed updating of bypass definitions
- Removed custom bypasses
- Removed Ace
- Changed how preferences are handled 
- Made BG script compatible with manifest version 3
- Re-wrote crowd bypass functions
- Broke Firefox

Note: Firefox has not implemented most of the mv3 API, a backported version will have to be made

Checklist:

- [x] I made sure there are no unnecessary changes in the code*
- [x] Tested on Chromium- Chrome & Edge Windows
- ❌ Tested on Firefox

### CODE REVIEW

File: scripts/build.js [hunk 1]
```diff
@@ -9,151 +9,198 @@
...
+    const ext_base_URL = script_src.searchParams.get("ext_base_URL")
 
-if (bypasses.hasOwnProperty(location.host)) {
+if (location.host in bypasses) {
     const bypass_url = bypasses[location.host];
     
-    import(\`\${window.x8675309bp}\${bypass_url}\`).then(({default: bypass}) => {
-        import(\`\${window.x8675309bp}helpers/dom.js\`).then(({default: helpers}) => {
+    
+    import(\`\${ext_base_URL}/\${bypass_url}\`).then(({default: bypass}) => {
```
Comment (reashetyrr) [inline scripts/build.js]: are we certain the match created above does not include the `/`? when i made it it had the`/` in the match....

File: src/js/injection_script.js [hunk 2]
```diff
@@ -1,63 +1,13 @@
...
+    const script_src = new URL(document.currentScript.src)
+    const ext_base_URL = script_src.searchParams.get("ext_base_URL")
```
Comment (reashetyrr) [inline src/js/injection_script.js]: this is useless, you need to add this to the generator, this file is generated

Comment (NotAProton) [reply src/js/injection_script.js]: yea that's prettier reformatting files, should I add injection script to gitignore ?https://softwareengineering.stackexchange.com/questions/391804/should-generated-documentation-be-stored-in-a-git-repository

File: scripts/build.js (hunk 1 above)
Comment (NotAProton) [reply scripts/build.js]: yup, 

> ```
> function getBaseURL() {
>   return "chrome-extension://" + chrome.runtime.id;
> }
> ```

https://github.com/FastForwardTeam/FastForward/pull/921/files/7bf15e5a8880d3a776eabe42ec3feed86b39f8b2#diff-87ec9444c8215c0d59aa4ec02b3b1e2204fe089c4db91b508befc4db657e5bebR13

File: src/js/injection_script.js (hunk 2 above)
Comment (NotAProton) [reply src/js/injection_script.js:3]: I get what you mean now

File: scripts/build.js (hunk 1 above)
Comment (NotAProton) [reply scripts/build.js]: I get what you mean now https://github.com/FastForwardTeam/FastForward/pull/921/commits/f34c0aa84beece7e0ee4d009d1447727ff5845c6

File: package.json
Comment (ghost) [inline package.json:1]: We should look into using Dependabot for the dependencies. I'll look into implementing it.

File: src/helpers/dom.js [hunk 3]
```diff
@@ -1,336 +1,292 @@
...
-            break
-        case (/(flux\.li)/.exec(target) || {}).input:
-            url += '&safe_in=25'
-            break
-    }
-    unsafelyAssign(url)
+  switch (
+    target //All values here have been tested using "Take me to destinations after 0 seconds."
+  ) {
+    case (/(krnl\.place|hugegames\.io)/.exec(target) || {}).input:
```
Comment (ghost) [inline src/helpers/dom.js:163]: We definitely should use the infoBox method to let people know that we don't bypass this website because it's video game cheating,

File: src/js/injection_script.js (hunk 2 above)
Comment (ghost) [reply src/js/injection_script.js]: Probably should, there's no need to keep it around if it changes every time build.js is ran.

**Review by ghost — CHANGES_REQUESTED** [review]
Get rid of bypasssites.md, it's outdated and not used in any capacity.

File: src/helpers/dom.js (hunk 3 above)
Comment (NotAProton) [reply src/helpers/dom.js:163]: That can be another PR, this one focuses on getting the extension working with MV3

**Review by ghost — APPROVED** [review]
